src/validation.cpp

```diff
@@ -3532,6 +3539,38 @@ static bool ContextualCheckBlockHeader(const CBlockHeader& block, BlockValidatio
             return state.Invalid(BlockValidationResult::BLOCK_INVALID_HEADER, strprintf("bad-version(0x%08x)", block.nVersion),
                                  strprintf("rejected nVersion=0x%08x block", block.nVersion));
 
+    if (!ContextualCheckBlockHeaderVolatile(block, state, params, pindexPrev)) return false;
+
+    return true;
+}
+
+/** Context-dependent validity checks, but rechecked in ConnectBlock().
+ *  Note that -reindex-chainstate skips the validation that happens here!
+ */
+static bool ContextualCheckBlockHeaderVolatile(const CBlockHeader& block, BlockValidationState& state, const CChainParams& params, const CBlockIndex* pindexPrev) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    const Consensus::Params& consensusParams = params.GetConsensus();
+
+    // Enforce MUST_SIGNAL status of deployments
+    for (int j = 0; j < (int)Consensus::MAX_VERSION_BITS_DEPLOYMENTS; ++j) {
+        Consensus::DeploymentPos deployment_pos = Consensus::DeploymentPos(j);
+        ThresholdState deployment_state = VersionBitsState(pindexPrev, consensusParams, deployment_pos, versionbitscache);
+        if (deployment_state == ThresholdState::MUST_SIGNAL) {
+            if ((block.nVersion & VersionBitsMask(consensusParams, deployment_pos)) == 0 || (block.nVersion & VERSIONBITS_TOP_MASK) != VERSIONBITS_TOP_BITS) {
+                BIP9Stats stats = VersionBitsStatistics(pindexPrev, consensusParams, deployment_pos);
+                if (stats.elapsed == stats.period) {
+                    // first block in new period
+                    stats.count = stats.elapsed = 0;
+                }
+                ++stats.elapsed; ++stats.count;
```

Comment:
I don't think the `++stats.count` there is correct -- if the final block in the MUST_SIGNAL period does not signal, and the total number of blocks signalling in the period is `threshold-1`, this will cause lockinontimeout=true nodes to transition from MUST_SIGNAL to LOCKED_IN (and later to ACTIVE) while lockinontimeout=false nodes transition from STARTED to FAILED.

Setting `N=107` instead of a random value and removing the `random.shuffle(bits)` in the test case should duplicate the failure.